astropy v1.3 adds new warnings: need to add them to the tests

For anyone desperately trying to optimise their Xmas wrapping paper usage, or eggnog distribution, with Sherpa, be aware that AstroPy v1.3 adds one (or more) new warnings which trip up our tests . The following is not guaranteed to be a complete list:

```
% python setup.py test
...

==================================== ERRORS ====================================
______ ERROR at teardown of test_save_arrays_nocols_FITS.test_save_arrays ______

    def fin():
        warnings = [w for w in recwarn.list
                    if type(w.message) not in known_warnings or not known(w)]
    
>       assert 0 == len(warnings)
E       assert 0 == 1
E        +  where 1 = len([<_pytest.recwarn.RecordedWarning object at 0x7f3dc3d68a20>])

sherpa/conftest.py:108: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: AstropyDeprecationWarning: "clobber" was deprecated in version 1.3 and will be removed in a future version. Use argument "overwrite" instead. [astropy.utils.decorators]
_______ ERROR at teardown of test_save_arrays_cols_FITS.test_save_arrays _______

    def fin():
        warnings = [w for w in recwarn.list
                    if type(w.message) not in known_warnings or not known(w)]
    
>       assert 0 == len(warnings)
E       assert 0 == 1
E        +  where 1 = len([<_pytest.recwarn.RecordedWarning object at 0x7f3dc0c5bd30>])

sherpa/conftest.py:108: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: AstropyDeprecationWarning: "clobber" was deprecated in version 1.3 and will be removed in a future version. Use argument "overwrite" instead. [astropy.utils.decorators]
_________________ ERROR at teardown of test_save_pha.testWrite _________________

    def fin():
        warnings = [w for w in recwarn.list
                    if type(w.message) not in known_warnings or not known(w)]
    
>       assert 0 == len(warnings)
E       assert 0 == 1
E        +  where 1 = len([<_pytest.recwarn.RecordedWarning object at 0x7f3dc4137518>])

sherpa/conftest.py:108: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: AstropyDeprecationWarning: "clobber" was deprecated in version 1.3 and will be removed in a future version. Use argument "overwrite" instead. [astropy.utils.decorators]
______________________ ERROR at teardown of test_success _______________________

    def fin():
        warnings = [w for w in recwarn.list
                    if type(w.message) not in known_warnings or not known(w)]
    
>       assert 0 == len(warnings)
E       assert 0 == 1
E        +  where 1 = len([<_pytest.recwarn.RecordedWarning object at 0x7f3dbf74dac8>])

sherpa/conftest.py:108: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: AstropyDeprecationWarning: "clobber" was deprecated in version 1.3 and will be removed in a future version. Use argument "overwrite" instead. [astropy.utils.decorators]
----------------------------------------------------------------------
Ran 4 tests in 0.322s

OK (skipped=2)
______________________ ERROR at teardown of test_failure _______________________

    def fin():
        warnings = [w for w in recwarn.list
                    if type(w.message) not in known_warnings or not known(w)]
    
>       assert 0 == len(warnings)
E       assert 0 == 1
E        +  where 1 = len([<_pytest.recwarn.RecordedWarning object at 0x7f3dbf6d13c8>])

sherpa/conftest.py:108: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: AstropyDeprecationWarning: "clobber" was deprecated in version 1.3 and will be removed in a future version. Use argument "overwrite" instead. [astropy.utils.decorators]
======================================================================
FAIL: test_failure (sherpa.astro.utils.smoke.SmokeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/djburke/sherpa/sherpa-base/sherpa/astro/utils/smoke.py", line 175, in test_failure
    self.fail("Requested to fail")
AssertionError: Requested to fail

----------------------------------------------------------------------
Ran 4 tests in 0.320s

FAILED (failures=1, skipped=1)
____________________ ERROR at teardown of test_fits_failure ____________________

    def fin():
        warnings = [w for w in recwarn.list
                    if type(w.message) not in known_warnings or not known(w)]
    
>       assert 0 == len(warnings)
E       assert 0 == 1
E        +  where 1 = len([<_pytest.recwarn.RecordedWarning object at 0x7f3dbf64a7f0>])

sherpa/conftest.py:108: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: AstropyDeprecationWarning: "clobber" was deprecated in version 1.3 and will be removed in a future version. Use argument "overwrite" instead. [astropy.utils.decorators]
----------------------------------------------------------------------
Ran 4 tests in 0.319s

OK (skipped=2)
___________________ ERROR at teardown of test_xspec_failure ____________________

    def fin():
        warnings = [w for w in recwarn.list
                    if type(w.message) not in known_warnings or not known(w)]
    
>       assert 0 == len(warnings)
E       assert 0 == 1
E        +  where 1 = len([<_pytest.recwarn.RecordedWarning object at 0x7f3dbf521780>])

sherpa/conftest.py:108: AssertionError
----------------------------- Captured stderr call -----------------------------
WARNING: AstropyDeprecationWarning: "clobber" was deprecated in version 1.3 and will be removed in a future version. Use argument "overwrite" instead. [astropy.utils.decorators]
----------------------------------------------------------------------
Ran 4 tests in 0.320s

OK (skipped=2)
=================================== FAILURES ===================================
_________________ test_89_issues.test_warnings_are_gone_arrays _________________

self = <test_io.test_89_issues testMethod=test_warnings_are_gone_arrays>

    @requires_fits
    def test_warnings_are_gone_arrays(self):
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always")
            with NamedTemporaryFile() as f:
                ui.load_arrays(1, [1,2,3], [4,5,6])
                ui.save_data(1, f.name, ascii=True, clobber=True)
            with NamedTemporaryFile() as f:
                ui.save_data(1, f.name, ascii=False, clobber=True)
>           assert len(w) == 0
E           AssertionError: assert 1 == 0
E            +  where 1 = len([<warnings.WarningMessage object at 0x7f3dc3cd3780>])

sherpa/astro/io/tests/test_io.py:50: AssertionError
__________________ test_89_issues.test_warnings_are_gone_pha ___________________

self = <test_io.test_89_issues testMethod=test_warnings_are_gone_pha>

    @requires_fits
    @requires_data
    def test_warnings_are_gone_pha(self):
        pha = self.make_path("3c273.pi")
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always")
            with NamedTemporaryFile() as f:
                ui.load_pha(pha)
                ui.save_data(1, f.name, ascii=False, clobber=True)
>           assert len(w) == 0
E           AssertionError: assert 1 == 0
E            +  where 1 = len([<warnings.WarningMessage object at 0x7f3dc344ca58>])

sherpa/astro/io/tests/test_io.py:61: AssertionError
----------------------------- Captured stdout call -----------------------------
WARNING: systematic errors were not found in file '/home/djburke/sherpa/sherpa-base/sherpa/../sherpa-test-data/sherpatest/3c273.pi'
statistical errors were found in file '/home/djburke/sherpa/sherpa-base/sherpa/../sherpa-test-data/sherpatest/3c273.pi' 
but not used; to use them, re-read with use_errors=True
read ARF file /home/djburke/sherpa/sherpa-base/sherpa/../sherpa-test-data/sherpatest/3c273.arf
read RMF file /home/djburke/sherpa/sherpa-base/sherpa/../sherpa-test-data/sherpatest/3c273.rmf
WARNING: systematic errors were not found in file '/home/djburke/sherpa/sherpa-base/sherpa/../sherpa-test-data/sherpatest/3c273_bg.pi'
statistical errors were found in file '/home/djburke/sherpa/sherpa-base/sherpa/../sherpa-test-data/sherpatest/3c273_bg.pi' 
but not used; to use them, re-read with use_errors=True
read background file /home/djburke/sherpa/sherpa-base/sherpa/../sherpa-test-data/sherpatest/3c273_bg.pi
============================ pytest-warning summary ============================
WC1 None passing a string to pytest.main() is deprecated, pass a list of arguments instead.
 2 failed, 1229 passed, 64 skipped, 1 pytest-warnings, 7 error in 87.04 seconds 
```

--- a/sherpa/astro/io/tests/test_io.py
+++ b/sherpa/astro/io/tests/test_io.py
@@ -1,5 +1,5 @@
 #
-#  Copyright (C) 2016  Smithsonian Astrophysical Observatory
+#  Copyright (C) 2016, 2017  Smithsonian Astrophysical Observatory
 #
 #
 #  This program is free software; you can redistribute it and/or modify
@@ -17,12 +17,11 @@
 #  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 #
 
-from sherpa.utils import SherpaTestCase, requires_data, requires_fits, requires_xspec
+from sherpa.utils import SherpaTestCase, requires_data, requires_fits, \
+    requires_xspec
 from sherpa.astro import ui
 
-from unittest import skipIf
 from tempfile import NamedTemporaryFile
-import warnings
 
 
 class test_89_issues(SherpaTestCase):
@@ -40,22 +39,18 @@ class test_89_issues(SherpaTestCase):
 
     @requires_fits
     def test_warnings_are_gone_arrays(self):
-        with warnings.catch_warnings(record=True) as w:
-            warnings.simplefilter("always")
-            with NamedTemporaryFile() as f:
-                ui.load_arrays(1, [1,2,3], [4,5,6])
-                ui.save_data(1, f.name, ascii=True, clobber=True)
-            with NamedTemporaryFile() as f:
-                ui.save_data(1, f.name, ascii=False, clobber=True)
-            assert len(w) == 0
+        ui.load_arrays(1, [1, 2, 3], [4, 5, 6])
+        #  We now have logic in conftest.py to catch white-listed warnings and fail on unexpected ones.
+        #  We just need to make any warnings bubble up, here and in the following test.
+        with NamedTemporaryFile() as f:
+            ui.save_data(1, f.name, ascii=True, clobber=True)
+        with NamedTemporaryFile() as f:
+            ui.save_data(1, f.name, ascii=False, clobber=True)
 
     @requires_fits
     @requires_data
     def test_warnings_are_gone_pha(self):
         pha = self.make_path("3c273.pi")
-        with warnings.catch_warnings(record=True) as w:
-            warnings.simplefilter("always")
-            with NamedTemporaryFile() as f:
-                ui.load_pha(pha)
-                ui.save_data(1, f.name, ascii=False, clobber=True)
-            assert len(w) == 0
+        ui.load_pha(pha)
+        with NamedTemporaryFile() as f:
+            ui.save_data(1, f.name, ascii=False, clobber=True)

--- a/sherpa/conftest.py
+++ b/sherpa/conftest.py
@@ -1,5 +1,5 @@
 #
-#  Copyright (C) 2016  Smithsonian Astrophysical Observatory
+#  Copyright (C) 2016, 2017  Smithsonian Astrophysical Observatory
 #
 #
 #  This program is free software; you can redistribute it and/or modify
@@ -30,6 +30,12 @@ try:  # Python 3
 except ImportError:  # Python 2
     import mock
 
+try:
+    from astropy.utils.exceptions import AstropyDeprecationWarning
+    have_astropy = True
+except ImportError:
+    have_astropy = False
+
 
 TEST_DATA_OPTION = "--test-data"
 
@@ -39,14 +45,18 @@ def pytest_addoption(parser):
                      help="Alternative location of test data files")
 
 
-# Whilelist of known warnings. One can associate different warning messages to the same warning class
+# Whilelist of known warnings. One can associate different warning messages
+# to the same warning class
 known_warnings = {
     DeprecationWarning:
         [
             r"unorderable dtypes.*",
             r"Non-string object detected for the array ordering.*",
             r"using a non-integer number instead of an integer will result in an error in the future",
-            r"Use load_xstable_model to load XSPEC table models"
+            r"Use load_xstable_model to load XSPEC table models",
+            #  This does not have to do with Sherpa and is coming from some versions of
+            #  jupyter_client
+            r"metadata .* was set from the constructor.*",
         ],
     UserWarning:
         [
@@ -75,6 +85,16 @@ if sys.version_info >= (3, 2):
             ]
     }
     known_warnings.update(python3_warnings)
+
+
+if have_astropy:
+    astropy_warnings = {
+        AstropyDeprecationWarning:
+        [
+            r".*clobber.*deprecated.*1.3",
+        ],
+    }
+    known_warnings.update(astropy_warnings)
 
 
 @pytest.fixture(scope="function", autouse=True)

